### PR TITLE
Initialization of conservative systems

### PR DESCRIPTION
**Pull request by wthrowe**
### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp [hunk 1]
```diff
@@ -85,151 +84,238 @@ namespace Actions {
...
+            next_time.with_slab(next_time.slab().advance_towards(initial_dt)));
       }
+
+      const size_t num_grid_points = db::get<Tags::Extents<Dim>>(box).product();
+      const auto& inertial_coords =
+          db::get<Tags::Coordinates<Tags::ElementMap<Dim>,
+                                    Tags::LogicalCoordinates<Dim>>>(box);
+
+      // Will be overwritten before use
+      DtVars dt_vars{num_grid_points, 0.0};
```
Comment (osheamonn) [inline src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp]: It the 0's will be overwritten, should it be initialised with signalling nans? Is this the default behavious for variables?

**Review by osheamonn — COMMENTED** [review]
I really like this and the changes to initialise element

File: src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp (hunk 1 above)
Comment (kidder) [reply src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp]: yes, without the `0.0`  it would be filled with signalling_nans

**Review by osheamonn — DISMISSED** [review]

**Comment by wthrowe** [comment]
There is a bug in the conservative initialization path.  Fixing it is easy, but the test is giving me difficulties so an update will have to wait for tomorrow.

**Review by kidder — COMMENTED** [review]
squash

**Review by kidder — DISMISSED** [review]
@nilsdeppe please look at

File: src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp [hunk 2]
```diff
@@ -180,12 +225,32 @@ struct InitializeElement {
         dt_variables_tag,
         Tags::HistoryEvolvedVariables<variables_tag, dt_variables_tag>>;
 
-    using compute_tags = db::AddComputeTags<
-        Tags::Time,
-        Tags::deriv<typename variables_tag::tags_list,
-                    typename System::gradients_tags,
-                    Tags::InverseJacobian<Tags::ElementMap<Dim>,
-                                          Tags::LogicalCoordinates<Dim>>>>;
+    template <typename LocalSystem, typename = std::nullptr_t>
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp]: I'd just make the second template parameter a bool so we don't need to do SFINAE, which is slower to compile

File: src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp [hunk 3]
```diff
@@ -114,8 +116,11 @@ struct InitializeElement {
   };
 
   // Tags related only to the system
+  template <typename System, typename = std::nullptr_t>
+  struct SystemTags;
+
   template <typename System>
-  struct SystemTags {
+  struct SystemTags<System, Requires<not System::is_conservative>> {
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp]: Use  a `bool` as the second template parameter would be faster than SFINAE. You should be able to default it to `System::is_conservative`

File: src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp [hunk 4]
```diff
@@ -252,41 +317,97 @@ struct InitializeElement {
...
+      return db::create_from<
+          db::RemoveTags<>,
+          db::AddSimpleTags<typename flux_comm_types::simple_mortar_data_tag,
+                            Tags::Mortars<Tags::Next<temporal_id_tag>, Dim>>>(
           std::move(box), std::move(mortar_data),
-          std::move(mortar_next_temporal_ids), std::move(normal_dot_fluxes));
+          std::move(mortar_next_temporal_ids));
+    }
+
+    template <typename LocalSystem, typename = std::nullptr_t>
```
Comment (nilsdeppe) [inline src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp]: Same bool comment again

**Review by nilsdeppe — COMMENTED** [review]
Minor suggestions, I'm happy for you to squash right away :)

**Review by kidder — DISMISSED** [review]
@nilsdeppe please look at

**Comment by wthrowe** [comment]
Rebased on the Mesh changes.

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**